transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -842,10 +843,8 @@ public ChannelFuture writeAndFlush(Object msg) {
     }
 
     private static void notifyOutboundHandlerException(Throwable cause, ChannelPromise promise) {
-        if (!promise.tryFailure(cause) && !(promise instanceof VoidChannelPromise)) {
-            if (logger.isWarnEnabled()) {
-                logger.warn("Failed to fail the promise because it's done already: {}", promise, cause);
-            }
+        if (!(promise instanceof VoidChannelPromise)) {
```

Comment:
for 4.0 what about adding an "unsable internal" public utility method `isVoid(ChannelPromise)` in the `io.netty.channel` package that can be used by `PromiseNotificationUtil`?